### PR TITLE
Issue 5211: Adds EventHelperFuture to fix NPE in StreamMetadataTasks

### LINKED ISSUE #5211: Flaky test StreamMetadataTest#testMetadataOperations on 0.8.0rc1 code
**Problem description**
While testing 0.8.0rc1 I saw this flaky test.
I am using JDK11 on Mac

i```
o.pravega.test.integration.controller.server.StreamMetadataTest > testMetadataOperations FAILED
    java.lang.NullPointerException
        at io.pravega.controller.task.Stream.StreamMetadataTasks.sealStream(StreamMetadataTasks.java:494)
        at io.pravega.controller.server.ControllerService.sealStream(ControllerService.java:224)
        at io.pravega.controller.server.eventProcessor.LocalController.sealStream(LocalController.java:212)
        at io.pravega.test.integration.controller.server.StreamMetadataTest.testMetadataOperations(StreamMetadataTest.java:93)
```

**Problem location**
The test

**Suggestions for an improvement**
Make the test consistently pass

### PR DESCRIPTION
**Pull request by kotlasaicharanreddy**
**Change log description**  

- Introduction of eventhelper in `StreamMetadataTasks.java` is not initialized in constructor but after the object after an explicit `initialize()` method.
- This causes `Null Pointer Exception` if request is received before initialize can happen.
- Added `EventHelperFuture` to fix NullPointerException.

**Purpose of the change**  
Fixes #5211

**What the code does**  

- Replaces `eventhelper` with `EventHelperFuture` accordingly. 

**How to verify it**  

- `testeventHelperNPE()` test should always pass.
- Build should succeed and all test must pass.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java [hunk 1]
```diff
@@ -135,15 +135,15 @@
     private static final int READER_GROUP_OPERATION_MAX_RETRIES = 10;
 
     private final AtomicLong retentionFrequencyMillis;
-    
+
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:138]: nit: revert please

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java [hunk 2]
```diff
@@ -1853,8 +1854,14 @@ public TaskBase copyWithContext(Context context) {
...
-        if (eventHelper != null) {
-            eventHelper.close();
+        if (eventHelperFuture.isDone()) {
+            eventHelperFuture.thenAccept(eventHelper -> {
+                if (eventHelper != null) {
+                    eventHelper.close();
+                }
+            }).join();
+        } else {
+            eventHelperFuture.cancel(true);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:1864]: This isn't going to have any effect. 

Furthermore, above (in `initialize`) when you set the value, that call will have no effect if this future is already completed (whether successfully or with exception). That means that the `EventHelper` object you create will be a source of resource leaks as you never close it.

Please rewrite your code to handle that situation.

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java [hunk 3]
```diff
@@ -190,17 +190,18 @@ private StreamMetadataTasks(final StreamMetadataStore streamMetadataStore,
...
         this.retentionClock = new AtomicReference<>(System::currentTimeMillis);
+        this.eventHelperFuture = new CompletableFuture<>();
         this.setReady();
     }
 
     @Synchronized
     public void initializeStreamWriters(final EventStreamClientFactory clientFactory,
                                         final String streamName) {
-        this.eventHelper = new EventHelper(clientFactory.createEventWriter(streamName,
+        this.eventHelperFuture.complete(new EventHelper(clientFactory.createEventWriter(streamName,
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:200]: This is a bad pattern. Never complete a future while holding a lock. You are holding the instance-level lock here with `@Synchronized`.

I recommend you rewrite this code to either remove the `@Synchronized` or complete this future outside of the scope of the lock.

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 2 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:1863]: we should do the following:
create eventHelper and set it in a lock protected variable and use a boolean to indicate if it needs to be set or not.
init method:
```
synchronize(lock) {
if(toSetEventHelper) {
eventhelper = new EventHelper();
toSetEventHelper = false;
}
}
// after we release the lock, if close was invoked then eventhelper may have been closed. but completing the future with that is safe (and its ok even if eventhelperfuture was cancelled in close method).
if (eventHelper != null) {
eventHelperFuture.complete(eventHelper);
}

```
close method:
```
synchronize(lock) {
toSetEventHelper = false;
if (eventhelper != null) {
eventHelper.close();
}
}
eventhelperfuture.cancel(true);
```

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 3 above)
Comment (kotlasaicharanreddy) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java]: Changes are done as per your recommendation.

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 2 above)
Comment (kotlasaicharanreddy) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java]: Changes are done as per suggested.

**Review by andreipaduroiu — APPROVED** [review]
My changes have been approved. @shiveshr it's in your hands now.

**Review by shiveshr — APPROVED** [review]
LGTM

**Pull request merged**